src/librustc_mir/transform/const_prop.rs

```diff
@@ -658,10 +657,64 @@ impl<'mir, 'tcx> ConstPropagator<'mir, 'tcx> {
         if rvalue.needs_subst() {
             return None;
         }
+        // self.use_ecx(|this| {
+        //     this.ecx.eval_rvalue_into_place(rvalue, place)
+        // })
+        self.eval_rvalue_with_identities(rvalue, place)
+    }
 
+    // Attempt to use albegraic identities to eliminate constant expressions
+    fn eval_rvalue_with_identities(
+        &mut self,
+        rvalue: &Rvalue<'tcx>,
+        place: Place<'tcx>,
+    ) -> Option<()> {
         self.use_ecx(|this| {
-            trace!("calling eval_rvalue_into_place(rvalue = {:?}, place = {:?})", rvalue, place);
-            this.ecx.eval_rvalue_into_place(rvalue, place)?;
+            match rvalue {
+                Rvalue::BinaryOp(op, left, right) | Rvalue::CheckedBinaryOp(op, left, right) => {
+                    let l = this.ecx.eval_operand(left, None);
+                    let r = this.ecx.eval_operand(right, None);
+
+                    if l.is_ok() && r.is_ok() {
```

Comment:
changing this line to `if let (Ok(_), Ok(_)) = (&l, &r)` causes failures in `crossbeam-utils`. I have NO idea why

Reply:
didn't CI just pass with the let version?

Reply:
nope! I changed it before hand. It would fail 100% of the time with `let` and pass 100% of the time with `is_ok()`

Reply:
that is scary. This is either a broken cache or a miscompilation, neither is good. Did you try `./x.py clean` before running again with the `let` variant?

Side note: This could all be rewritten as a big match:

```rust
let const_arg = match (l, r) {
    (Ok(x), Err(_)) | (Err(_), Ok(x)) => x,
    (Err(e), Err(_)) => return Err(e),
    (Ok(_), Ok(_)) => {
        this.ecx.eval_rvalue_into_place(rvalue, place)?;
        return Ok(());
    },
};
```

Reply:
yea I did actually try ` ./x.py clean` because I accidentally broke my build cache trying to add debug statements in the deps.

Reply:
Hmm I just tried to reproduce this issue again, and it is no longer occuring?